include/RequestLimiter.inc.php

```diff
@@ -0,0 +1,198 @@
+    local filled_tokens = math.min(capacity, last_tokens+(delta*rate))
+    local allowed = filled_tokens >= requested
+    local new_tokens = filled_tokens
+    if allowed then
+      new_tokens = filled_tokens - requested
+    end
+    
+    redis.call("setex", tokens_key, ttl, new_tokens)
+    redis.call("setex", timestamp_key, ttl, now)
+    
+    return { allowed, new_tokens }';
+
+	protected $concurrent_requests_limiter_lua = '
+    local key = KEYS[1]
+
+    local capacity = tonumber(ARGV[1])
+    local timestamp = tonumber(ARGV[2])
+    local id = ARGV[3]
+    
+    local count = redis.call("zcard", key)
+    local allowed = count < capacity
+    
+    if allowed then
+      redis.call("zadd", key, timestamp, id)
+    end
+    
+    return { allowed, count }';
+
+	public function __construct() {
+		$this->redis = Z_Redis::get();
+		$this->request_rate_limiter_lua_sha1 = sha1($this->request_rate_limiter_lua); //we can hardcode sha1 to prevent calculating it every time
+		$this->concurrent_requests_limiter_lua_sha1 = sha1($this->concurrent_requests_limiter_lua);
+	}
+
+	/**
+	 * Limit request rate for a given bucket
+	 * @param $params - bucket (userid or key), limit (requests per second), burst (multiplier for 'limit'), warn (threshold for remaining requests warning)
+	 * @return array|null
+	 */
+	public function limitRate($params) {
+		$bucket = $params['bucket'];
+		$rate = $params['limit'];
+		$burst = $params['burst'];
+
+		$capacity = $burst * $rate;
+
+		$prefix = 'rrl:' . $bucket;
+
+		$keys = [$prefix . '.tokens', $prefix . '.timestamp'];
+
+		$args = [$rate, $capacity, time(), 1];
+
+		try {
+			$res = $this->redis->evalSha($this->request_rate_limiter_lua_sha1, array_merge($keys, $args), count($keys));
+			if (!$res) {
+				Z_Core::logError('Executing evalSha failed in Z_RequestLimiter::limitRate, maybe sha1 is wrong');
+
+				$res = $this->redis->eval($this->request_rate_limiter_lua, array_merge($keys, $args), count($keys));
+
+				if (!$res) {
+					Z_Core::logError('Executing eval failed in Z_RequestLimiter::limitRate');
+					return null;
+				}
+			}
+		} catch (Exception $e) {
+			Z_Core::logError('Redis exception in Z_RequestLimiter::limitRate: ' . $e->getMessage());
+			return null;
+		}
+
+		return [
+			'allowed' => $res[0],
+			'bucket' => $params['bucket'],
+			'remaining' => $res[1],
+			'low' => $params['warn'] >= $res[1]
+		];
+	}
+
+	/**
+	 * Limit concurrent request for a bucket.
+	 * This function must be stared before actual API request logic.
+	 * finishConcurrent must be called every time after finishing the API request logic.
+	 * @param $params - bucket (userid or key), ttl (seconds), limit (requests per second)
+	 * @return array|null
+	 */
+	public function beginConcurrent($params) {
+		$ttl = $params['ttl']; //seconds how long the token will be kept (if not removed by finishConcurrent)
+		$capacity = $params['limit'];
+		$timestamp = time();
+		$id = Zotero_Utilities::randomString(10, 'mixed');
```

Comment:
I haven't done the math, and not that it matters much, but 10 is probably excessive here, right?